lib/galaxy/webapps/galaxy/api/users.py

```diff
@@ -240,3 +250,353 @@ def anon_user_api_value( self, trans ):
+            token = payload.get('token')
+            token_result = None
+            if token:
+                # If a token was supplied, validate and set user
+                token_result = trans.sa_session.query(trans.app.model.PasswordResetToken).get(token)
+                if not token_result or not token_result.expiration_time > datetime.utcnow():
+                    raise MessageException('Invalid or expired password reset token, please request a new one.')
+                user = token_result.user
+            else:
+                # The user is changing their own password, validate their current password
+                user = self._get_user(trans, id)
+                (ok, message) = trans.app.auth_manager.check_change_password(user, current)
+                if not ok:
+                    raise MessageException(message)
+            if user:
+                # Validate the new password
+                message = validate_password(trans, password, confirm)
+                if message:
+                    raise MessageException(message)
+                else:
+                    # Save new password
+                    user.set_password_cleartext(password)
+                    # if we used a token, invalidate it and log the user in.
+                    if token_result:
+                        trans.handle_user_login(token_result.user)
+                        token_result.expiration_time = datetime.utcnow()
+                        trans.sa_session.add(token_result)
+                    # Invalidate all other sessions
+                    for other_galaxy_session in trans.sa_session.query(trans.app.model.GalaxySession) \
+                                                     .filter(and_(trans.app.model.GalaxySession.table.c.user_id == user.id,
+                                                                  trans.app.model.GalaxySession.table.c.is_valid == true(),
+                                                                  trans.app.model.GalaxySession.table.c.id != trans.galaxy_session.id)):
+                        other_galaxy_session.is_valid = False
+                        trans.sa_session.add(other_galaxy_session)
+                    trans.sa_session.add(user)
+                    trans.sa_session.flush()
+                    trans.log_event('User change password')
+                    return {'message': 'Password has been saved.'}
+            raise MessageException('Failed to determine user, access denied.')
+        else:
+            return {'message': 'Password unchanged.',
+                    'inputs': [ {'name': 'current', 'type': 'password', 'label': 'Current password'},
+                                {'name': 'password', 'type': 'password', 'label': 'New password'},
+                                {'name': 'confirm', 'type': 'password', 'label': 'Confirm password'},
+                                {'name': 'token', 'type': 'hidden', 'hidden': True, 'ignore': None} ]}
+
```

Comment:
Thanks @jmchilton. Both are excellent comments. That the logout endpoint assumes a state definitely needs to be addressed. Since this PR is already pretty big and introduces a currently disabled feature, it would be fine to remove and leave it out for now. Regarding returning full input dictionaries, I generally agree with your argument. In fact many of the earlier commits build some (less dynamic) forms in the front-end and others in the backend, sometimes mixing both. In some cases separation is easy. The password form you mention above is a perfect example. With other forms it lead to inconsistencies. The filters e.g. are pretty dynamic in such that the backend decides which filters are enabled, what their format is and in which categories they ought to be displayed, similar is true for permissions. Decoupling them enabled each API endpoint to respond in arbitrary ways and lead to custom build front-end modules to compensate for that i.e. each API point sends an individually formatted dictionary (with pretty much the same information) and then each front-end module converts those into the established 'input' dictionaries. A consequence of that is that each API endpoint has to be re-learned. The tool form builder e.g. responds with exactly the same input dictionaries. Having said this I know that it is a compromise when API responses contain something like 'label' or 'hidden' and in this case even more so since there is no underlying tool.xml and one could statically implement these attributes (or even entire forms) in the front-end. In Charts e.g. everything is build in the front-end and stays there since there is no API endpoint which evaluates or validates the posted form state changes, requiring knowledge of the input types. I think this is something for a hangout session, so we can figure out what should be where. Eitherway its very feasible to move things around and we can totally do that. I will share the hangout link with @martenson and others once we find a good time for it.